### PR TITLE
AS4 and Fastify v4 Update: Better align API with Fastify conventions and general maintenance

### PR DESCRIPTION
**Pull request by olyop**
Draft pull request to go along side #6635 where changes have be outlined there.

In summary:
1. Update dependencies for Fastify v4 while still supporting v3. 
2. Changed connecting API to better align with Fastify plugin conventions and make use of the `fastify-plugin` package.
3. Change existing internal handlers and functions to use async/await.
4. Also updating the package for upcoming AS4 release.

### CODE REVIEW

File: packages/apollo-server-fastify/package.json [hunk 1]
```diff
@@ -26,16 +26,16 @@
     "node": ">=12.0"
   },
   "dependencies": {
-    "@fastify/accepts": "^3.0.0",
-    "@fastify/cors": "^7.0.0",
+    "@fastify/accepts": "^4.0.0",
+    "@fastify/cors": "^8.0.0",
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply packages/apollo-server-fastify/package.json:30]: I got this error when running tests:
`FastifyError: fastify-plugin: @fastify/accepts - expected '4.x' fastify version, '3.29.1' is installed`

File: packages/apollo-server-fastify/package.json [hunk 2]
```diff
@@ -26,16 +26,16 @@
...
     "apollo-server-types": "file:../apollo-server-types"
   },
   "devDependencies": {
     "apollo-server-integration-testsuite": "file:../apollo-server-integration-testsuite"
   },
   "peerDependencies": {
-    "fastify": "^3.17.0",
-    "graphql": "^15.3.0 || ^16.0.0"
+    "fastify": "^3.29.0 || ^4.2.0",
+    "graphql": "^15.8.0 || ^16.5.0"
```
Comment (trevor-scheer) [inline packages/apollo-server-fastify/package.json:39]: We can't narrow the peer dependency range without a major version bump, so we need to keep the lower bounds as they were:
```suggestion
    "fastify": "^3.17.0 || ^4.2.0",
    "graphql": "^15.3.0 || ^16.5.0"
```

And ideally, we should support a whole major version when we expand (so the 4.2.0 would be 4.0.0 and 16.5.0 would be 16.0.0).

File: packages/apollo-server-fastify/src/ApolloServer.ts [hunk 3]
```diff
@@ -24,134 +26,171 @@ export interface FastifyContext {
...
+      pluginOptions,
+    );
+
+  async #createGraphQLServerOptions(
     request: FastifyRequest,
     reply: FastifyReply,
   ): Promise<GraphQLOptions> {
     const contextParams: FastifyContext = { request, reply };
-    return this.graphQLServerOptions(contextParams);
+    return super.graphQLServerOptions(contextParams)
```
Comment (trevor-scheer) [inline packages/apollo-server-fastify/src/ApolloServer.ts:50]: I can't run tests with these changes, something about `super` in a private method (using `#`) is causing a syntax error. If I remove the `#` or revert back to `this` I don't see that error.

1) Can we just keep using `this`? Why the change to `super`?
2) I'm not opposed to using `#` as a private modifier, but the convention in the codebase is to use TS's `private`.

**Comment by olyop** [comment]
It looks like this pull request can only be an update for Fastify v4 as the packages `@fastify/cors` and `@fastify/accepts` only support v4 in their [plugin metadata](https://github.com/fastify/fastify-accepts/blob/master/index.js#L60). It wouldn't be backwards compatible with fastify v3 so it's one or the other.

This is unfortunate as these plugins definitely can support both versions they just don't specify it.

We could either update this package now to 4.x only supporting Fastify v4 and the current version for Fastify v3 or we could wait until AS4. For the first option that would mean there would be another major update when AS4 comes out which is soon.

Once the docs are updated I think these changes could be merged if everyone is happy with it but I will get rid of the deprecation warnings and add them back in the next release.

**Comment by trevor-scheer** [comment]
@olyop I'm not convinced supporting 3 and 4 is a good idea for this package. If this package were structured with a better use of peerDependencies, this might be possible, but changing that now would be breaking. Making tweaks to be more permissive at runtime like what you suggested seems very unsafe - if those packages were compatible with both v3 and v4 wouldn't they have specified that?

From what I've seen so far and based on our constraints (no breaking changes), Fastify v4 is not an option for `apollo-server-fastify@3`. I'm happy for us to migrate to the more conventional way of doing things if that's possible within Fastify v3. We will not be publishing an `apollo-server-fastify@4` which is compatible with `apollo-server@3`.

My hope for the AS4 work is a new package entirely, not owned by the `apollographql` org at all (you are welcome to create an org for it or it can simply belong to your account if you like). We're happy to support integrations in AS4 but don't intend to own them if there's community interest in maintaining them. To me, this is the ideal place to make the switch to Fastify v4 and make breaking changes as you see fit.

If anything is unclear please let me know. I don't want to stonewall your great work on this, but as I mentioned before, breaking changes are a nonstarter, even if that means we're stuck doing things the old way on this current version. AS4+ is promising for this very reason! The future Fastify package will be free to make breaking changes and major version upgrades as it makes sense rather than be locked the the current Apollo Server versioning lifecycle.

**Comment by trevor-scheer** [comment]
Hey @olyop, your email said you'd incorporated my latest feedback but I only see a merge commit from main. Did you forget to push maybe?

**Comment by glasser** [comment]
This PR has been replaced by @olyop 's work on `@as-integrations/fastify`!

**Pull request closed without merging**